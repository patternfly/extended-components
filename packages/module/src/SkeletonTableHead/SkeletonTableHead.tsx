import React, { ReactNode, useMemo } from 'react';
import {
  Th,
  ThProps,
  Thead,
  Tr
} from '@patternfly/react-table';
import { Skeleton } from '@patternfly/react-core';

export const isThObject = (value: ReactNode | { cell: ReactNode; props?: ThProps }): value is { cell: ReactNode; props?: ThProps } => value != null && typeof value === 'object' && 'cell' in value;

export interface SkeletonTableHeadProps {
  /** Custom columns for the table */
  columns?: (ReactNode | { cell: ReactNode; props?: ThProps })[];
  /** Number of columns in the table */
  columnsCount?: number;
  /** Custom OUIA ID */
  ouiaId?: string | number;
  /** Flag indicating if the table is selectable */
  isSelectable?: boolean;
  /** Flag indicating if the table is expandable */
  isExpandable?: boolean;
  /** Flag indicating if the table is a tree table */
  isTreeTable?: boolean;
}

export const SkeletonTableHead: React.FunctionComponent<SkeletonTableHeadProps> = ({
  ouiaId = 'SkeletonTableHeader',
  isSelectable,
  isExpandable,
  columnsCount,
  columns,
  isTreeTable,
  ...rest
}: SkeletonTableHeadProps) => {
  const hasCustomColumns = Array.isArray(columns);
  const rowCellsCount = hasCustomColumns ? columns.length : columnsCount;

  const cells = useMemo(() => [
    ...(isExpandable ? [ <Th key="row-expand" screenReaderText='Data expansion table header cell' /> ] : []),
    ...(isSelectable && !isTreeTable ? [ <Th key="row-select" screenReaderText='Data selection table header cell' /> ] : []),
    ...(hasCustomColumns ? (
      columns.map((column, index) => (
        <Th key={index} {...(isThObject(column) && column?.props)} data-ouia-component-id={`${ouiaId}-th-${index}`}>
          {isThObject(column) ? column.cell : column}
        </Th>
      ))
    ) : (
      [ ...Array(rowCellsCount) ].map((_, index) => (
        <Th key={index} data-ouia-component-id={`${ouiaId}-th-${index}`}>
          <Skeleton />
        </Th>
      ))
    )) 
  ]
  , [ hasCustomColumns, isExpandable, isSelectable, isTreeTable, columns, rowCellsCount, ouiaId ]);
  

  return (
    <Thead data-ouia-component-id={`${ouiaId}-thead`} {...rest}>
      <Tr ouiaId={`${ouiaId}-tr-head`}>
        {cells}
      </Tr>
    </Thead>
  );
};

export default SkeletonTableHead;
