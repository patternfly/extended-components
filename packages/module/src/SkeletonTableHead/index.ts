export { default } from './SkeletonTableHead';
export * from './SkeletonTableHead';
