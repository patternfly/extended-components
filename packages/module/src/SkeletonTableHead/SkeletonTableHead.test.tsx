import React from 'react';
import { render } from '@testing-library/react';
import SkeletonTableHead from './SkeletonTableHead';

describe('SkeletonTableHead component', () => {
  it('should render correctly with count', () => {
    expect(render(<SkeletonTableHead columnsCount={2} isSelectable isExpandable />)).toMatchSnapshot();
  });

  it('should render correctly with count', () => {
    expect(render(<SkeletonTableHead columns={[ 'First', 'Second' ]} isTreeTable isSelectable />)).toMatchSnapshot();
  });
});
