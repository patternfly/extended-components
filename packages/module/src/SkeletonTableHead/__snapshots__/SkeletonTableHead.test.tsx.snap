// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SkeletonTableHead component should render correctly with count 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <thead
        class="pf-v5-c-table__thead"
        data-ouia-component-id="SkeletonTableHeader-thead"
      >
        <tr
          class="pf-v5-c-table__tr"
          data-ouia-component-id="SkeletonTableHeader-tr-head"
          data-ouia-component-type="PF5/TableRow"
          data-ouia-safe="true"
        >
          <th
            class="pf-v5-c-table__th"
            scope="col"
            tabindex="-1"
          >
            <span
              class="pf-v5-screen-reader"
            >
              Data expansion table header cell
            </span>
          </th>
          <th
            class="pf-v5-c-table__th"
            scope="col"
            tabindex="-1"
          >
            <span
              class="pf-v5-screen-reader"
            >
              Data selection table header cell
            </span>
          </th>
          <th
            class="pf-v5-c-table__th"
            data-ouia-component-id="SkeletonTableHeader-th-0"
            scope="col"
            tabindex="-1"
          >
            <div
              class="pf-v5-c-skeleton"
            >
              <span
                class="pf-v5-screen-reader"
              />
            </div>
          </th>
          <th
            class="pf-v5-c-table__th"
            data-ouia-component-id="SkeletonTableHeader-th-1"
            scope="col"
            tabindex="-1"
          >
            <div
              class="pf-v5-c-skeleton"
            >
              <span
                class="pf-v5-screen-reader"
              />
            </div>
          </th>
        </tr>
      </thead>
    </div>
  </body>,
  "container": <div>
    <thead
      class="pf-v5-c-table__thead"
      data-ouia-component-id="SkeletonTableHeader-thead"
    >
      <tr
        class="pf-v5-c-table__tr"
        data-ouia-component-id="SkeletonTableHeader-tr-head"
        data-ouia-component-type="PF5/TableRow"
        data-ouia-safe="true"
      >
        <th
          class="pf-v5-c-table__th"
          scope="col"
          tabindex="-1"
        >
          <span
            class="pf-v5-screen-reader"
          >
            Data expansion table header cell
          </span>
        </th>
        <th
          class="pf-v5-c-table__th"
          scope="col"
          tabindex="-1"
        >
          <span
            class="pf-v5-screen-reader"
          >
            Data selection table header cell
          </span>
        </th>
        <th
          class="pf-v5-c-table__th"
          data-ouia-component-id="SkeletonTableHeader-th-0"
          scope="col"
          tabindex="-1"
        >
          <div
            class="pf-v5-c-skeleton"
          >
            <span
              class="pf-v5-screen-reader"
            />
          </div>
        </th>
        <th
          class="pf-v5-c-table__th"
          data-ouia-component-id="SkeletonTableHeader-th-1"
          scope="col"
          tabindex="-1"
        >
          <div
            class="pf-v5-c-skeleton"
          >
            <span
              class="pf-v5-screen-reader"
            />
          </div>
        </th>
      </tr>
    </thead>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`SkeletonTableHead component should render correctly with count 2`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <thead
        class="pf-v5-c-table__thead"
        data-ouia-component-id="SkeletonTableHeader-thead"
      >
        <tr
          class="pf-v5-c-table__tr"
          data-ouia-component-id="SkeletonTableHeader-tr-head"
          data-ouia-component-type="PF5/TableRow"
          data-ouia-safe="true"
        >
          <th
            class="pf-v5-c-table__th"
            data-ouia-component-id="SkeletonTableHeader-th-0"
            scope="col"
            tabindex="-1"
          >
            First
          </th>
          <th
            class="pf-v5-c-table__th"
            data-ouia-component-id="SkeletonTableHeader-th-1"
            scope="col"
            tabindex="-1"
          >
            Second
          </th>
        </tr>
      </thead>
    </div>
  </body>,
  "container": <div>
    <thead
      class="pf-v5-c-table__thead"
      data-ouia-component-id="SkeletonTableHeader-thead"
    >
      <tr
        class="pf-v5-c-table__tr"
        data-ouia-component-id="SkeletonTableHeader-tr-head"
        data-ouia-component-type="PF5/TableRow"
        data-ouia-safe="true"
      >
        <th
          class="pf-v5-c-table__th"
          data-ouia-component-id="SkeletonTableHeader-th-0"
          scope="col"
          tabindex="-1"
        >
          First
        </th>
        <th
          class="pf-v5-c-table__th"
          data-ouia-component-id="SkeletonTableHeader-th-1"
          scope="col"
          tabindex="-1"
        >
          Second
        </th>
      </tr>
    </thead>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
